lms/djangoapps/instructor/views/api.py

```diff
@@ -726,6 +728,343 @@ def extract_user_info(user):
+@cache_control(no_cache=True, no_store=True, must_revalidate=True)
+@require_level('instructor')
+def delete_saved_query(_request, course_id, query_to_delete):  # pylint: disable=unused-argument
+    """
+    Deletes a grouped query that the user has saved along with the corresponding subqueries
+    """
+    data_access.delete_saved_query(query_to_delete)
+    response_payload = {
+        'success': True
+    }
+    return JsonResponse(response_payload)
+
+
+@ensure_csrf_cookie
+@cache_control(no_cache=True, no_store=True, must_revalidate=True)
+@require_level('instructor')
+def save_query(request, course_id):
+    """
+    Saves a group of queries and assigns them the same group ID
+    """
+    existing = request.GET.get('existing')
+    if existing is not None and existing != "":
+        existing_queries = existing.split(',')
+    else:
+        existing_queries = []
+    course_id = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+    clean_existing = [query for query in existing_queries if (query != "working" and query != "")]
+    success = data_access.save_query(course_id, clean_existing)
+
+    response_payload = {
+        'course_id': course_id.to_deprecated_string(),
+        'success': success
+    }
+    return JsonResponse(response_payload)
+
+
+@ensure_csrf_cookie
+@cache_control(no_cache=True, no_store=True, must_revalidate=True)
+@require_level('instructor')
+def get_temp_queries(_request, course_id):
+    """
+    Returns the temporary user queries per course
+    """
+    course_id = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+    saved_temp = data_access.get_temp_queries(course_id)
+    cleaned_queries = []
+    for query in saved_temp:
+        cleaned_queries.append({'id': query.id,
+                                'inclusion': REVERSE_INCLUSION_MAP[query.inclusion],
+                                'block_type': query.module_state_key.block_type,
+                                'block_id': query.module_state_key.block_id,
+                                'filter_on': query.filter_on,
+                                'display_name': query.entity_name,
+                                'type': query.type,
+                                'done': query.done,
+                                })
+    response_payload = {
+        'course_id': course_id.to_deprecated_string(),
+        'queries': cleaned_queries
+    }
+    return JsonResponse(response_payload)
+
+
+@ensure_csrf_cookie
+@cache_control(no_cache=True, no_store=True, must_revalidate=True)
+@require_level('instructor')
+def get_saved_queries(_request, course_id):
+    """
+    Returns all the user-saved queries per course
+    """
+    course_id = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+    groups, queries, relations = data_access.get_saved_queries(course_id)
+    cleaned_queries = []
+    relation_map = {}
+    created = {}
+    for relation in relations:
+        relation_map[relation.query_id] = relation.grouped_id
```

Comment:
This can be converted to a dictionary comprehension like:

```python
relation_map = {
    relation.query_id: relation.grouped_id
    for relation in relations
}
```